File: sudoku.py
"""
Sudoku solver with backtracking search and AC-3 algorithm
"""

import copy

class Sudoku:
    def __init__(self, grid):
        """Initialize the variables and domains based on the given grid"""
        # Raise exception if the grid of the sudoku is not 9x9
        if len(grid) != 9 or not all(len(row) == 9 for row in grid):
            raise Exception("Sudoku grid must be 9x9.")

        self.explored = 0

        # Get the variables which are the empty cells in the grid
        self.variables = []
        for i in range(9):
            for j in range(9):
                if not grid[i][j]:
                    self.variables.append((i, j))

        # Assign the value of the empty cells with None initially
        self.values = grid

        # Assign domains to variables based on unary constraint to ensure node consistency
        self.domains = copy.deepcopy(grid)
        for cell in self.variables:
            x, y = cell
            self.domains[x][y] = set(range(1, 10))
            # Unary constraint on row
            for i in range(9):
                if i == y:
                    continue
                if grid[x][i] in self.domains[x][y]:
                    self.domains[x][y].remove(grid[x][i])

            # Unary constraint on column
            for i in range(9):
                if i == x:
                    continue
                if grid[i][y] in self.domains[x][y]:
                    self.domains[x][y].remove(grid[i][y])

            # Unary constraint on sub-grid
            start_x = x - (x % 3)
            start_y = y - (y % 3)

            for i in range(start_x, start_x + 3):
                for j in range(start_y, start_y + 3):
                    if i == x and j == y:
                        continue
                    if grid[i][j] in self.domains[x][y]:
                        self.domains[x][y].remove(grid[i][j])

    def solve(self, show_explored=False):
        """Returns a 2D list of solution"""
        solution = self.backtrack(self.values, self.domains)
        explored = self.explored

        if show_explored:
            return (solution, explored)
        else:
            return solution

    def backtrack(self, values, domains):
        """Solve the puzzle by using backtracking search. Recursively calling itself until there is a solution"""
        # Check whether the puzzle has been solved
        if all(None not in row for row in values):
            return values

        self.explored += 1

        # Select an unassigned variable
        var = self.unassigned_variable(values, domains)
        x, y = var

        # Iterate over all values in the domain of the cell
        for value in domains[x][y]:
            new_values = copy.deepcopy(values)
            new_domains = copy.deepcopy(domains)
            new_values[x][y] = value
            # Check whether the new value assigned is arc consistent
            if self.consistent(var, new_values):
                new_domains[x][y] = {value}
                # Draw inferences based on the new value assigned
                inferences = self.ac3(var, new_domains)
                # Consider next option if the cell has no solution
                if not inferences:
                    continue
                # Call itself recursively to consider options for other cells
                result = self.backtrack(new_values, new_domains)
                if result:
                    return result
        return None


    def unassigned_variable(self, values, domains):
        """Select an unassigned variable using Minimum Remaining Values (MRV) and Degree heuristic"""
        # Initialize the value of MRV which is greater than the maximum possible value
        mrv = 10
        
        # Initialize the minimum possible value of degree
        degree = 0

        candidate = None

        unassigned_variables = []
        for i in range(9):
            for j in range(9):
                if values[i][j] is None:
                    unassigned_variables.append((i, j))

        for var in unassigned_variables:
            x, y = var
            if len(domains[x][y]) <= mrv:
                candidate = var
                mrv = len(domains[x][y])
                if len(self.neighbours(var)) > degree:
                    candidate = var
                    degree = len(self.neighbours(var))

        return candidate

    def consistent(self, var, values):
        """Check arc consistency for binary constraint"""
        for neighbour in self.neighbours(var):
            x, y = var
            i, j = neighbour
            if values[x][y] == values[i][j]:
                return False

        return True

    def neighbours(self, var):
        """Returns neighbours for a given cell"""
        x, y = var

        neighbours = set()
        # Neighbours for same row
        for i in range(9):
            if not i == y and (x, i) in self.variables:
                neighbours.add((x, i))

        # Neighbours for same column
        for i in range(9):
            if not i == x and (i, y) in self.variables:
                neighbours.add((i, y))

        # Neighbours for same sub-grid
        start_x = x - (x % 3)
        start_y = y - (y % 3)

        for i in range(start_x, start_x + 3):
            for j in range(start_y, start_y + 3):
                if (not (i == x and j == y)) and (i, j) in self.variables:
                    neighbours.add((i, j))

        return list(neighbours)

    def ac3(self, var, domains):
        """Ensure arc consistency for all related nodes of the variable"""
        queue = []
        for neighbour in self.neighbours(var):
            queue.append((neighbour, var))
        while queue:
            a, b = queue.pop(0)
            if self.revise(a, b, domains):
                # No possible solution
                if len(domains[a[0]][a[1]]) == 0:
                    return False
                new_neighbours = list(set(self.neighbours(a)) - {b})
                for c in new_neighbours:
                    # New inferences only can be made if it is the only value in the domain 
                    if len(domains[a[0]][a[1]]) == 1:
                        queue.append((c, a))
        return True

    def revise(self, a, b, domains):
        """Ensure arc consistency between two nodes"""
        revised = False
        duplicate_values = []
        for value in domains[a[0]][a[1]]:
            # Remove duplicate values if it is inconsistent between two nodes
            if value in domains[b[0]][b[1]]:
                duplicate_values.append(value)
                revised = True
        for value in duplicate_values:
            domains[a[0]][a[1]].remove(value)
        return revised

File: javascript/sudoku.js
/**
 * Sudoku solver with backtracking search and AC-3 algorithm
 */

class Sudoku {
    constructor(grid) {
        /**
         * Initialise the variables and domains based on the given grid
         */

        // Throw exception if the grid of the sudoku is not 9x9
        if (grid.length !== 9) throw "Sudoku grid must be 9x9.";
        
        for (let i = 0; i < 9; i++) {
            if (grid[i].length !== 9) throw "Sudoku grid must be 9x9.";
        }
        
        this.explored = 0;

        // Get the variables which are the empty cells in the grid
        this.variables = [];
        for (let i = 0; i < 9; i++) {
            for (let j = 0; j < 9; j++) {
                if (!grid[i][j]) this.variables.push([i, j]);
            }
        }

        // Assign the value of the empty cells with None initially
        this.values = grid;

        // Assign domains to variables based on unary constraint to ensure node consistency
        this.domains = structuredClone(grid);
        for (let cell of this.variables) {
            let [x, y] = cell;
            this.domains[x][y] = new Set([1, 2, 3, 4, 5, 6, 7, 8, 9])

            // Unary constraint on row
            for (let i = 0; i < 9; i++) {
                if (i === y) continue;
                if (this.domains[x][y].has(grid[x][i])) this.domains[x][y].delete(grid[x][i]);
            }

            // Unary constraint on column
            for (let i = 0; i < 9; i++) {
                if (i === x) continue;
                if (this.domains[x][y].has(grid[i][y])) this.domains[x][y].delete(grid[i][y]);
            }

            // Unary constraint on sub-grid
            let start_x = x - (x % 3);
            let start_y = y - (y % 3);

            for (let i = start_x; i < start_x + 3; i++) {
                for (let j = start_y; j < start_y + 3; j++) {
                    if (i === x && j === y) continue;
                    if (this.domains[x][y].has(grid[i][j])) this.domains[x][y].delete(grid[i][j]);
                }
            }
        }
    }
    
    solve(show_explored=false) {
        /**
         * Returns a 2D list of solution
         */
        let solution = this.backtrack(this.values, this.domains)
        let explored = this.explored

        return (show_explored) ? [solution, explored] : solution
    }

    backtrack(values, domains) {
        /**
         * Solve the puzzle by using backtracking search. Recursively calling itself until there is a solution
         */

        // Check whether the puzzle has been solved
        if (values.every(row => row.every(cell => cell !== null))) return values;

        this.explored++;

        // Select an unassigned variable
        let v = this.unassigned_variable(values, domains);
        let [x, y] = v;

        // Iterate over all values in the domain of the cell
        for (let value of domains[x][y]) {
            let new_values = structuredClone(values);
            let new_domains = structuredClone(domains);
            new_values[x][y] = value;
            // usestate here; new values
            // Check whether the new value assigned is arc consistent
            if (this.consistent(v, new_values)) {
                new_domains[x][y] = new Set([value]);
                // Draw inferences based on the new value assigned
                let inferences = this.ac3(v, new_domains);
                // Consider next option if the cell has no solution
                if (!inferences) continue;
                // Call itself recursively to consider options for other cells
                let result = this.backtrack(new_values, new_domains);
                if (result) return result;
            }
        }
        return null;
    }

    unassigned_variable(values, domains) {
        /**
         * Select an unassigned variable using Minimum Remaining Values (MRV) and Degree heuristic
         */

        // Initialise the value of MRV which is greater than the maximum possible value
        let mrv = 10
        
        // Initialise the minimum possible value of degree
        let degree = 0;

        let candidate = null;

        let unassgined_variables = []
        for (let i = 0; i < 9; i++) {
            for (let j = 0; j < 9; j++) {
                if (values[i][j] === null) unassgined_variables.push([i, j]);
            }
        }

        for (let v of unassgined_variables) {
            let [x, y] = v;
            if (domains[x][y].size <= mrv) {
                candidate = v;
                mrv = domains[x][y].size;
                let n = this.neighbours(v).length;
                if (n > degree) {
                    candidate = v;
                    degree = n;
                }
            }
        }
        return candidate;
    }

    consistent (v, values) {
        /**
         * Check arc consistency for binary constraint
         */

        for (let neighbour of this.neighbours(v)) {
            let [x, y] = v;
            let [i, j] = neighbour;
            if (values[x][y] === values[i][j]) return false;
        }
        return true;
    }

    neighbours (v) {
        /**
         * Returns neighbours for a given cell
         */

        let [x, y] = v;

        let neighbours = new Set();

        // Neighbours for same row
        for (let i = 0; i < 9; i++) {
            if (!(i === y) && (itemInArray(this.variables, [x, i]))) neighbours.add(JSON.stringify([x, i]));
        }

        // Neighbours for same column
        for (let i = 0; i < 9; i++) {
            if (!(i === x) && (itemInArray(this.variables, [i, y]))) neighbours.add(JSON.stringify([i, y]));
        }

        // Neighbours for same sub-grid
        let start_x = x - (x % 3);
        let start_y = y - (y % 3);

        for (let i = start_x; i < start_x + 3; i++) {
            for (let j = start_y; j < start_y + 3; j++) {
                if (!((i === x) && (j === y)) && (itemInArray(this.variables, [i, j]))) neighbours.add(JSON.stringify([i, j]));
            }
        }

        let neighbours_array = []
        for (let elem of neighbours) {
            neighbours_array.push(JSON.parse(elem));
        }
        return neighbours_array;
    }

    ac3(v, domains) {
        /**
         * Ensure arc consistency for all related nodes of the variable
         */

        let queue = [];
        for (let neighbour of this.neighbours(v)) {
            queue.push([neighbour, v]);
        }

        while (queue.length > 0) {
            let [a, b] = queue.shift();
            if (this.revise(a, b, domains)) {
                // No possible solution
                if (domains[a[0]][b[0]].size === 0) return false;
                let s = new Set(this.neighbours(a));
                let new_neighbours = Array.from(s).filter(function(elem) {
                    return (JSON.stringify(elem) !== JSON.stringify(b))
                });
                for (let c of new_neighbours) {
                    // New inferences only can be made if it is the only value in the domain
                    if (domains[a[0]][a[1]].size === 1) queue.push([c, a]);
                }
            }
        }
        return true;
    }

    revise(a, b, domains) {
        /**
         * Ensure arc consistency between two nodes
         */
        for (let value of domains[a[0]][a[1]]) {
            // Remove duplicate values if it is inconsistent between two nodes
            if (domains[b[0]][b[1]].has(value)) {
                domains[a[0]][a[1]].delete(value);
                return true;
            }
        }
        return false;
    }
}

// Check if an array is in another array
function itemInArray(arr, item) {
    return arr.some(elem => {
        return JSON.stringify(elem) === JSON.stringify(item);
    });
}

export default Sudoku;
